File: Parallel-Java/src/parallel/linq101/mylinq/AverageMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Order;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;

public class AverageMethods {
    //original:4 target:40
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }
    //average syntax : 5

    public double numbersAverage() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).average().getAsDouble();
        return averageNum;
    }

    public double bigNumbersAverage() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).filter(n -> n >= 5).average().getAsDouble();
        return averageNum;
    }

    public double oddNumbersAverage() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).average().getAsDouble();
        return averageNum;
    }

    public double doubleAverage() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double averageNum = Arrays.stream(numbers).average().getAsDouble();
        return averageNum;
    }

    public double bigDoubleNumbersAverage() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double averageNum = Arrays.stream(numbers).filter(n -> n >= 5).average().getAsDouble();
        return averageNum;
    }


    //average projection: 15

    public double numbersAverageTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).map(n -> n * 2).average().getAsDouble();
        return averageNum;
    }

    public double bigNumbersAverageTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).filter(n -> n >= 5).map(n -> n * 2).average().getAsDouble();
        return averageNum;
    }

    public double oddNumbersAverageTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        double averageNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).map(n -> n * 2).average().getAsDouble();
        return averageNum;
    }

    public double stringAverageLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        double averageLength = Arrays.stream(words).mapToInt(w -> w.length()).average().getAsDouble();
        return averageLength;
    }

    public double shortStringAverageLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        double averageLength = Arrays.stream(words).filter(w -> w.length() < 7).mapToInt(w -> w.length()).average().getAsDouble();
        return averageLength;
    }

    public double productUnitPriceAverage() {
        List<Product> products = getProductList();
        double averageUnitPrice = products.stream().mapToDouble(p -> p.getUnitPrice()).average().getAsDouble();
        return averageUnitPrice;
    }

    public double expensiveProductUnitPriceAverage() {
        List<Product> products = getProductList();
        double averageUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToDouble(p -> p.getUnitPrice()).average().getAsDouble();
        return averageUnitPrice;
    }

    public double expensiveProductStockAverage() {
        List<Product> products = getProductList();
        double averageUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToInt(p -> p.getUnitsInStock()).average().getAsDouble();
        return averageUnitPrice;
    }

    public double productLongNameLengthAverage() {
        List<Product> products = getProductList();
        double averageUnitPrice = products.stream().map(p -> p.getProductName())
                .filter(name -> name.length() > 7).mapToInt(name -> name.length()).average().getAsDouble();
        return averageUnitPrice;
    }

    public List<Pair<String, Double>> customerOrdersTotalAverage() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalAverage = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().mapToDouble(o -> o.getTotal()).average().getAsDouble())).toList();
        return ordersTotalAverage;
    }

    public Map<String, Double> customerOrdersTotalAverageDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalAverage = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).average().getAsDouble()));
        return ordersTotalAverage;
    }

    public List<Pair<String, Double>> customerInRegionOrdersTotalAverage(String region) {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalAverage = customers.stream().filter(c -> c.getRegion().equals(region))
                .map(c -> Pair.with(c.getCustomerId(),
                        c.getOrders().stream().mapToDouble(o -> o.getTotal()).average().getAsDouble())).toList();
        return ordersTotalAverage;
    }

    public Map<String, Double> customerInRegionOrdersTotalAverageDict(String region) {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalAverage = customers.stream().filter(c -> c.getRegion().equals(region))
                .collect(toMap(c -> c.getCustomerId(),
                        c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).average().getAsDouble()));
        return ordersTotalAverage;
    }

    public List<Pair<String, Double>> customerExpensiveOrdersTotalAverage() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalAverage = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).average().getAsDouble())).toList();
        return ordersTotalAverage;
    }

    public Map<String, Double> customerExpensiveOrdersTotalAverageDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalAverage = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).average().getAsDouble()));
        return ordersTotalAverage;
    }

    //average grouped: 10

    public List<Pair<String, Double>> productEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Double>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).average().getAsDouble())).toList();
        return categories;
    }

    public Map<String, Double> productEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Double> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).average().getAsDouble()));
        return categories;
    }

    public List<Pair<String, Double>> expensiveProductEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Double>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).average().getAsDouble())).toList();
        return categories;
    }

    public Map<String, Double> expensiveProductEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Double> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).average().getAsDouble()));
        return categories;
    }

    public List<Pair<String, Double>> customerEachRegionOrdersAverage() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersAverage = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).average().getAsDouble())).toList();
        return ordersAverage;
    }

    public Map<String, Double> customerEachRegionOrdersAverageDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersAverage = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).average().getAsDouble()));
        return ordersAverage;
    }

    public List<Pair<String, Double>> customerEachCountryOrdersAverage() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersAverage = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).average().getAsDouble())).toList();
        return ordersAverage;
    }

    public Map<String, Double> customerEachCountryOrdersAverageDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersAverage = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).average().getAsDouble()));
        return ordersAverage;
    }

    public List<Pair<String, Double>> customerEachRegionOrdersTotalAverage() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalAverage = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).average().getAsDouble())).toList();
        return ordersTotalAverage;
    }

    public Map<String, Double> customerEachRegionOrdersTotalAverageDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalAverage = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).average().getAsDouble()));
        return ordersTotalAverage;
    }
}

File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/AverageMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class AverageMethod {
    //original:3 target:30
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    //average syntax : 5

    public double NumbersAverage() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Average();
        return averageNum;
    }

    public double BigNumbersAverage() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Where(n => n > 5).Average();
        return averageNum;
    }

    public double OddNumbersAverage() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Where(n => n % 2 == 1).Average();
        return averageNum;
    }

    public double DoubleAverage() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double averageNum = numbers.Average();
        return averageNum;
    }

    public double BigDoubleNumbersAverage() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double averageNum = numbers.Where(d => d > 5).Average();
        return averageNum;
    }


    //average projection: 15

    public double NumbersAverageTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Average(n => n * 2);
        return averageNum;
    }

    public double BigNumbersAverageTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Where(n => n > 5).Average(n => n * 2);
        return averageNum;
    }

    public double OddNumbersAverageTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        double averageNum = numbers.Where(n => n % 2 == 1).Average(n => n * 2);
        return averageNum;
    }

    public double StringAverageLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        double averageLength = words.Average(w => w.Length);
        return averageLength;
    }

    public double ShortStringAverageLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        double averageLength = words.Where(w => w.Length < 7).Average(w => w.Length);
        return averageLength;
    }

    public double ProductUnitPriceAverage() {
        List<Product> products = GetProductList();
        double averageUnitPrice = products.Average(p => p.UnitPrice);
        return averageUnitPrice;
    }

    public double ExpensiveProductUnitPriceAverage() {
        List<Product> products = GetProductList();
        double averageUnitPrice = products.Where(p => p.UnitPrice > 100).Average(p => p.UnitPrice);
        return averageUnitPrice;
    }

    public double ExpensiveProductStockAverage() {
        List<Product> products = GetProductList();
        double averageUnitPrice = products.Where(p => p.UnitPrice > 100).Average(p => p.UnitsInStock);
        return averageUnitPrice;
    }

    public double ProductLongNameLengthAverage() {
        List<Product> products = GetProductList();
        double averageUnitPrice = products.Select(p => p.ProductName)
            .Where(name => name.Length > 7).Average(name => name.Length);
        return averageUnitPrice;
    }

    public List<Tuple<string, double>> CustomerOrdersTotalAverage() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalAverage = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Average(o => o.Total))).ToList();
        return ordersTotalAverage;
    }

    public Dictionary<string, double> CustomerOrdersTotalAverageDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalAverage = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Average(o => o.Total));
        return ordersTotalAverage;
    }

    public List<Tuple<string, double>> CustomerInRegionOrdersTotalAverage(string region) {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalAverage = customers.Where(c => c.Region.Equals(region))
            .Select(c => Tuple.Create(c.CustomerId,
                c.Orders.Average(o => o.Total))).ToList();
        return ordersTotalAverage;
    }

    public Dictionary<string, double> CustomerInRegionOrdersTotalAverageDict(string region) {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalAverage = customers.Where(c => c.Region.Equals(region))
            .ToDictionary(c => c.CustomerId,
                c => c.Orders.Average(o => o.Total));
        return ordersTotalAverage;
    }

    public List<Tuple<string, double>> CustomerExpensiveOrdersTotalAverage() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalAverage = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Where(o => o.Total > 100).Average(o => o.Total))).ToList();
        return ordersTotalAverage;
    }

    public Dictionary<string, double> CustomerExpensiveOrdersTotalAverageDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalAverage = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Where(o => o.Total > 100).Average(o => o.Total));
        return ordersTotalAverage;
    }

    //average grouped: 10

    public List<Tuple<string, double>> ProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, double>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Average(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, double> ProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, double> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Average(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, double>> ExpensiveProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, double>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Where(p => p.UnitPrice > 100)
                .Average(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, double> ExpensiveProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, double> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g =>
                g.Where(p => p.UnitPrice > 100).Average(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, double>> CustomerEachRegionOrdersAverage() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersAverage = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.Average(c => c.Orders.Count))).ToList();
        return ordersAverage;
    }

    public Dictionary<string, double> CustomerEachRegionOrdersAverageDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersAverage = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.Average(c => c.Orders.Count));
        return ordersAverage;
    }

    public List<Tuple<string, double>> CustomerEachCountryOrdersAverage() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersAverage = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Average(c => c.Orders.Count))).ToList();
        return ordersAverage;
    }

    public Dictionary<string, double> CustomerEachCountryOrdersAverageDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersAverage = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Average(c => c.Orders.Count));
        return ordersAverage;
    }

    public List<Tuple<string, double>> CustomerEachRegionOrdersTotalAverage() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalAverage = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c => c.Orders).Average(o => o.Total))).ToList();
        return ordersTotalAverage;
    }

    public Dictionary<string, double> CustomerEachRegionOrdersTotalAverageDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalAverage = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.SelectMany(c => c.Orders).Average(o => o.Total));
        return ordersTotalAverage;
    }
}